### PR TITLE
Build dnlib on UNIX-based OS

### LINKED ISSUE #246: dnlib cannot be build on UNIX systems
Despite .NET Core support, the project cannot be build on UNIX systems due to the implicit reference to full framework within the targetframeworks property.

### PR DESCRIPTION
**Pull request by twsI**
Possible fix for #246.

Closes #246

### CODE REVIEW

**Comment by 0xd4d** [comment]
This completely disables building anything but the netcoreapp target on Linux/macOS. @kiootic added some code to be able to build it on Linux if mono is installed, see this: https://github.com/0xd4d/dnlib/blob/master/src/dnlib.csproj#L69-L76

Is it not possible for you to use the NuGet package?

**Comment by 0xd4d** [comment]
If you update `Examples.csproj` with netcoreapp2.0 -> netcoreapp2.1 it should be possible to build it from the root dir without any other changes: `dotnet build -c Release -f netcoreapp2.1`

**Comment by twsI** [comment]
Since there is no nightly nuget package, I decided to stick to the submodule. I simply missed the mono support section, but I'll look into it and try to come up with a solution that satisfies everyone. Feel free to cherry pick the first commit for now.

**Comment by twsI** [comment]
I believe, I referece the target frameworks correctly now. Unfortunately I haven't been able to verify the result with mono, xbuild complained about the project format [missing some XML](https://github.com/Microsoft/msbuild/pull/1089). .NET Core on Win/Ubuntu and via VS works fine.

**Comment by 0xd4d** [comment]
> xbuild complained about the project format [missing some XML](https://github.com/Microsoft/msbuild/pull/1089).

The latest mono has msbuild, so xbuild shouldn't be needed.

@kiootic Do you still need to build it with mono? Or can the mono hacks be removed?

**Comment by kiootic** [comment]
Personally I don't work with C# frequently anymore, and the projects I involved in had migrated to using .NET Standard. So I don't mind its removal.

File: Examples/Examples.csproj [hunk 1]
```diff
@@ -1,7 +1,8 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>net461;netcoreapp2.0</TargetFrameworks>
+    <TargetFrameworks>netcoreapp2.1</TargetFrameworks>
+    <TargetFrameworks Condition=" '$(MSBuildRuntimeType)' != 'Core' OR '$(OS)' == 'Windows_NT' ">$(TargetFrameworks);net461</TargetFrameworks>
```
Comment (0xd4d) [inline Examples/Examples.csproj]: It's should be enough to just check the OS here.

File: src/dnlib.csproj [hunk 2]
```diff
@@ -1,15 +1,16 @@
...
+﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <Import Project="$(SolutionDir)\DnlibCommon.props" Condition="Exists('$(SolutionDir)\DnlibCommon.props')" />
 
   <PropertyGroup>
     <DefineConstants>$(DefineConstants);$(MoreDefineConstants)</DefineConstants>
     <DefineConstants Condition=" '$(DnlibThreadSafe)' != 'false' ">$(DefineConstants);THREAD_SAFE</DefineConstants>
-    <TargetFrameworks>netstandard2.0;net461;netcoreapp2.1</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0;netcoreapp2.1</TargetFrameworks>
+    <TargetFrameworks Condition=" '$(MSBuildRuntimeType)' != 'Core' OR '$(OS)' == 'Windows_NT' ">$(TargetFrameworks);net461</TargetFrameworks>
```
Comment (0xd4d) [inline src/dnlib.csproj]: Same here

File: src/dnlib.csproj [hunk 3]
```diff
@@ -1,15 +1,16 @@
...
   <PropertyGroup>
     <DefineConstants>$(DefineConstants);$(MoreDefineConstants)</DefineConstants>
     <DefineConstants Condition=" '$(DnlibThreadSafe)' != 'false' ">$(DefineConstants);THREAD_SAFE</DefineConstants>
-    <TargetFrameworks>netstandard2.0;net461;netcoreapp2.1</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0;netcoreapp2.1</TargetFrameworks>
+    <TargetFrameworks Condition=" '$(MSBuildRuntimeType)' != 'Core' OR '$(OS)' == 'Windows_NT' ">$(TargetFrameworks);net461</TargetFrameworks>    
     <!-- It's not possible to target .NET Framework 3.5 when using dotnet build
          https://github.com/Microsoft/msbuild/issues/1333 -->
     <NoTargetFrameworkNet35 Condition=" '$(NoTargetFrameworkNet35)' == '' AND '$(MSBuildRuntimeType)' == 'Core' ">true</NoTargetFrameworkNet35>
-    <TargetFrameworks Condition=" '$(NoTargetFrameworkNet35)' != 'true' ">$(TargetFrameworks);net35</TargetFrameworks>
```
Comment (0xd4d) [inline src/dnlib.csproj]: Could you restore the space?

File: src/dnlib.csproj [hunk 4]
```diff
@@ -66,7 +67,7 @@ For better *Windows PDB* writer support, you should add a reference to `Microsof
     <DefineConstants>$(DefineConstants);TRACE</DefineConstants>
   </PropertyGroup>
 
-  <PropertyGroup Condition=" '$(OS)' != 'Windows_NT' ">
```
Comment (0xd4d) [inline src/dnlib.csproj]: This whole block can be removed now, we don't need to support building with mono.

**Comment by twsI** [comment]
Everything should be fixed now. I completly removed explicit mono support as you proposed.

File: src/dnlib.csproj [hunk 5]
```diff
@@ -1,15 +1,16 @@
...
   <PropertyGroup>
     <DefineConstants>$(DefineConstants);$(MoreDefineConstants)</DefineConstants>
     <DefineConstants Condition=" '$(DnlibThreadSafe)' != 'false' ">$(DefineConstants);THREAD_SAFE</DefineConstants>
-    <TargetFrameworks>netstandard2.0;net461;netcoreapp2.1</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0;netcoreapp2.1</TargetFrameworks>
+    <TargetFrameworks Condition=" '$(OS)' == 'Windows_NT' ">$(TargetFrameworks);net461</TargetFrameworks>
     <!-- It's not possible to target .NET Framework 3.5 when using dotnet build
          https://github.com/Microsoft/msbuild/issues/1333 -->
     <NoTargetFrameworkNet35 Condition=" '$(NoTargetFrameworkNet35)' == '' AND '$(MSBuildRuntimeType)' == 'Core' ">true</NoTargetFrameworkNet35>
-    <TargetFrameworks Condition=" '$(NoTargetFrameworkNet35)' != 'true' ">$(TargetFrameworks);net35</TargetFrameworks>
```
Comment (0xd4d) [inline src/dnlib.csproj]: This was not fixed

File: src/dnlib.csproj [hunk 6]
```diff
@@ -32,10 +33,11 @@
     <IncludeSymbols>true</IncludeSymbols>
     <SymbolPackageFormat>snupkg</SymbolPackageFormat>
 
-    <PackageDescription>Reads and writes .NET assemblies and modules, Windows PDBs and Portable PDBs.
```
Comment (0xd4d) [inline src/dnlib.csproj]: Undo the formatting. Whatever tool you're using, it does things it shouldn't be doing.

Comment (twsI) [reply src/dnlib.csproj]: This is Visual Studio 15.9.7 default formatting

**Comment by 0xd4d** [comment]
Thanks. VS is wrong about that formatting though, only leading and trailing spaces are removed, but it thought it was OK to indent text in the middle of the description.

**Pull request merged**